lib/src/solver/package_lister.dart

```diff
@@ -36,6 +36,15 @@ class PackageLister {
   /// solve isn't a `pub get`.
   final PackageId _locked;
 
+  // The version of this package that, if retracted, is still allowed in the
+  // current version solve.
+  //
+  // We don't allow a retracted version during solving unless it is already
+  // present in `pubspec.lock` or pinned in `dependency_overrides`.
+  //
+  // This is `null` if there is no retracted version that can be allowed.
+  final Version allowedRetractedVersion;
```

Comment:
I have a feeling we might want `_locked` and `allowedRetractedVersion` should be the same type.

Reply:
I have the same feeling, but I think it might be `_locked` that needs to change. I'll add to the cleanup I want to do in a separate PR.